### PR TITLE
Support arbitrary multi probe and multi trigger structures in spikeglx

### LINKED ISSUE #1149: SpikeGLX converter should support arbitrary multi probe and multi trigger structure
We have an example with multiple probes in the training data but can't be opened with the converter:

```python
from neuroconv.converters import SpikeGLXConverterPipe
from spikeinterface.extractors import SpikeGLXRecordingExtractor

folder_path = ECEPHY_DATA_PATH / "spikeglx"/ "multi_trigger_multi_gate" / "SpikeGLX" / "5-19-2022-CI0" 
assert folder_path.exists()

from neuroconv.converters import SpikeGLXConverterPipe

converter = SpikeGLXConverterPipe(folder_path=folder_path)
nwbfile = converter.create_nwbfile()


ValidationError: 1 validation error for __init__
file_path
  Path does not point to a file [type=path_not_file, input_value=PosixPath('/home/heberto/...22-CI0_t0.imec0.ap.bin'), input_type=PosixPath]"
}

```

The problem is caused by the fact that we have hard-code the file paths here:
1) We hardcode the path of the files here:
https://github.com/catalystneuro/neuroconv/blob/c4afad37a8600aed3c21fa73ce45fbb24ba82ea4/src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxconverter.py#L66-L85

Because this works just fine with the recording object on spikeinterface:

image

My plan is to first match the paths as proposed on https://github.com/catalystneuro/neuroconv/issues/978 so the problem in https://github.com/catalystneuro/neuroconv/issues/1112 can be addressed.

### PR DESCRIPTION
**Pull request by h-mayorquin**
This should fix #1149 and #978

### CODE REVIEW

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxconverter.py [hunk 1]
```diff
@@ -61,28 +63,17 @@ def __init__(
         """
         folder_path = Path(folder_path)
 
-        streams = streams or self.get_streams(folder_path=folder_path)
+        streams_ids = streams or self.get_streams(folder_path=folder_path)
 
         data_interfaces = dict()
-        for stream in streams:
```
Comment (h-mayorquin) [inline src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxconverter.py:67]: This is the critical part of this. Previously, the Converter assumed a fix structure of the files. This PR gets rids of that assumption.

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py [hunk 2]
```diff
@@ -54,38 +53,58 @@ def _source_data_to_extractor_kwargs(self, source_data: dict) -> dict:
     @validate_call
     def __init__(
         self,
-        file_path: FilePath,
-        verbose: bool = True,
+        folder_path: Optional[DirectoryPath] = None,
+        stream_id: Optional[str] = None,
         es_key: Optional[str] = None,
+        verbose: bool = True,
+        file_path: Optional[FilePath] = None,
```
Comment (h-mayorquin) [inline src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py:60]: I would like to deprecate this but I don't want this discussion to get in the way of this PR. If people who review this agree we could start a soft deprecation as we usually do.

Reminder: Spiekinterface takes folder path as input (as in this PR) and not file_path as we currently do in main.

File: tests/test_on_data/ecephys/test_spikeglx_converter.py [hunk 3]
```diff
@@ -193,7 +194,7 @@ def test_electrode_table_writing(tmp_path):
     np.testing.assert_array_equal(saved_channel_names, expected_channel_names_nidq)
 
     # Test AP
-    electrical_series = nwbfile.acquisition["ElectricalSeriesAPImec0"]
+    electrical_series = nwbfile.acquisition["ElectricalSeriesAP"]
```
Comment (h-mayorquin) [inline tests/test_on_data/ecephys/test_spikeglx_converter.py:197]: Previously, the Converter always writes the ElectricalSeries using the name `f"ElectricalSeriesAPImec{probe_index}"` whereas the interface uses the name `ElectricalSeriesAP` only. This PR makes the behavior of the converter and the interface consistent in the single probe case hence the change.

**Comment by h-mayorquin** [comment]
The failing test is unrelated to this and should be fixed by #1151

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,11 +6,13 @@
...
 * Fix a bug where data in `DeepLabCutInterface` failed to write when `ndx-pose` was not imported. [#1144](https://github.com/catalystneuro/neuroconv/pull/1144)
+* `SpikeGLXConverterPipe` converter now accepts multi-probe structures with multi-trigger and does not assume a specific folder structure [#1150](https://github.com/catalystneuro/neuroconv/pull/1150)
 
 ## Features
 * Propagate the `unit_electrode_indices` argument from the spikeinterface tools to `BaseSortingExtractorInterface`. This allows users to map units to the electrode table when adding sorting data [PR #1124](https://github.com/catalystneuro/neuroconv/pull/1124)
 * Imaging interfaces have a new conversion option `always_write_timestamps` that can be used to force writing timestamps even if neuroconv's heuristics indicates regular sampling rate [PR #1125](https://github.com/catalystneuro/neuroconv/pull/1125)
 * Added .csv support to DeepLabCutInterface [PR #1140](https://github.com/catalystneuro/neuroconv/pull/1140)
+* `SpikeGLXRecordingInterface` now also accepts `folder_path` making its behavior requivalent to SpikeInterface [#1150](https://github.com/catalystneuro/neuroconv/pull/1150)
```
Comment (bendichter) [inline CHANGELOG.md:15]:
```suggestion
* `SpikeGLXRecordingInterface` now also accepts `folder_path` making its behavior equivalent to SpikeInterface [#1150](https://github.com/catalystneuro/neuroconv/pull/1150)
```

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py (hunk 2 above)
Comment (bendichter) [inline src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py:60]: Let's preserve the order of the args so we do not cause a breaking change for ordered args

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxnidqinterface.py [hunk 5]
```diff
@@ -29,15 +30,15 @@ def get_source_schema(cls) -> dict:
...
         extractor_kwargs["folder_path"] = self.folder_path
         extractor_kwargs["stream_id"] = self.stream_id
         return extractor_kwargs
 
     @validate_call(config=ConfigDict(arbitrary_types_allowed=True))
     def __init__(
         self,
-        file_path: FilePath,
+        folder_path: Optional[DirectoryPath] = None,
+        file_path: Optional[FilePath] = None,
```
Comment (bendichter) [inline src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxnidqinterface.py:41]: same here, let's keep file_path as the first arg for now

**Review by bendichter — APPROVED** [review]

**Review by bendichter — CHANGES_REQUESTED** [review]
minor fixes

File: src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py [hunk 6]
```diff
@@ -54,38 +53,58 @@ def _source_data_to_extractor_kwargs(self, source_data: dict) -> dict:
...
-            else:
-                raise ValueError("Cannot automatically determine es_key from path")
-        file_path = Path(file_path)
-        self.folder_path = file_path.parent
+        if file_path is not None and stream_id is None:
+            self.stream_id = fetch_stream_id_for_spikelgx_file(file_path)
+            self.folder_path = Path(file_path).parent
+
+        else:
+            self.stream_id = stream_id
```
[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py:81]: SpikeInterface already throws an error that is similar, example:

image

Does this work or would you like something more specific?

[the remark this answers is not in the document]
Comment (bendichter) [reply src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py:81]: ok great! Yeah that works

[the remark this answers is not in the document]
Comment (h-mayorquin) [reply src/neuroconv/datainterfaces/ecephys/spikeglx/spikeglxdatainterface.py:81]: Anything else missing to merge thigs?

**Review by bendichter — APPROVED** [review]

**Pull request merged**